Explorer revamp

sal-uva opened the pull request:
This PR revamps how the Explorer works and looks. It specifically does the following:

- Adds a new `OPTION_DATASOURCES_TABLE` user input that creates a table with dynamic columns for each enabled dataset. Input fields per row can be text, dropdown, and checkbox
- Uses this table to create a new Settings page where the Explorer can be enabled per data source (more table options can be added later).
- Simplifies how custom data source templates for the Explorer are handled: they are now composed of CSS files (in `static/css/explorer/`) and Jinja2 templates (in `webtool/templates/explorer/datasource-templates/`) instead of CSS and JSON files in the data source folders that need to be verified and parsed.
- Integrates the Explorer with the UI of 4CAT.
- Makes the Explorer use `iterate_items`.
- Re-integrate sorting so that _all_ dataset columns can be used for sorting in the Explorer.
- Enable new annotation columns for for sorting, filtering, and other features.
- Adds Explorer templates for Twitter and Instagram (other data sources will soon follow).
- Deletes much unnecessary code.

Note that some unused code is still present for future updates with respect to 4CAT scrapers and database-accessible data sources generally.

dale-wahl reviewed (commented):
I looked over the backend code and only noticed one real issue (in an edge case). I ran this version and tested out the Explorer on a number of datasets (instagram, custom, telegram, tumblr, tiktok, youtube, reddit). It looks good! Sort works well. Reddit was missing the "subject" field (it's probably the only dataset that uses subject anymore). Telegram has an issue which I will post separately.

I tested saving annotations and writing them to datasets. This worked for me (and broke one with my edge case 😬; see comment). I did notice that the new fields show up in the Dataset preview view, but the values saved to the database do not show up in preview. The values do show up after you have run "write annotations".

Changing deactivating/activating settings seem to work fine. There is an `explorerflask` settings group that could probably be merged with the Explorer group.

If you want to merge now, I would deactivate Telegram as a default (till addressed) and consider how to address my comment re: field names for annotations.

sal-uva commented on datasources/ninegag/search_9gag.py:
```diff
@@ -8,6 +8,7 @@
 
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
True, we should do some cleanup..

sal-uva commented on processors/filtering/write_annotations.py:
```diff
@@ -101,7 +102,7 @@ def process(self):
 		# Write to top dataset
 		for label, values in new_data.items():
-			self.add_field_to_parent("annotation_" + label, values, which_parent=self.source_dataset, update_existing=True)
+			self.add_field_to_parent(label, values, which_parent=self.source_dataset, update_existing=True)
```
This is indeed an oversight for now, though I would like to have the option for annotation fields to have a 'clean' name; long names are quickly unreadable in spreadsheet software. This can be resolved by initially checking whether an annotation field key already exists in the dataset columns _or_, when annotated datasets are filtered and create a new dataset, if it is not a field registered in the `annotations` table for the parent dataset.

This is a sort-of edge case for now, but I'll try to resolve this next week!

sal-uva commented:
Since merging to master is no longer of immediate concern, I will address the above and make further improvements; will notify when it's time for a review!

dale-wahl commented on backend/lib/processor.py:
```diff
@@ -418,7 +418,7 @@ def add_field_to_parent(self, field_name, new_data, which_parent=source_dataset,
 
 		if len(new_data) != which_parent.num_rows:
-			raise ProcessorException('Must have new data point for each record: parent dataset: %i, new data points: %i' % (which_parent.num_rows, len(new_data)))
+			self.dataset.update_status('The amount of new data points and existing records don\'t match; data may be misaligned (parent dataset: %i, new data points: %i)' % (which_parent.num_rows, len(new_data)))
```
The issue was that there is no check ensuring new data is added to the correct row (for example, a check that the `id` is the same). If a row is missing, the function will naively add the data to whatever new row is at that _index_. It was raising an error to avoid adding the data to the wrong row/item.

stijn-uva replied:
Since this method is now no longer used for annotations or the explorer as far as I can see, but only for other processors, I would revert the changes made to it

sal-uva commented:
Okay, I think this is a good base version.

Changelog:
- A custom template can now no longer be chosen manually, but will automatically be used if template files are present.
- Post bodies can now be 'social mediafied' so tags, @-mentions, and URLs get an anchor tag that leads to the original social media site.
- Improved the Twitter, Instagram, and TikTok templates.
  - Twitter now shows contents of quoted tweets.
  - Instagram shows co-authors.
- Moved the functions that save annotations and annotation fields to `dataset.py`, so that they are useable by other processors in the future.
- You can now reverse-sort by the original dataset order.
- Removed quite some non-necessary code.
- Animations in showing/hiding annotations and the annotation fields editor.
- Misc bugfixes.

Still to do:
- Add templates for other data sources (Telegram, Tumblr, Reddit, LinkedIn, etc.).
- More rigorously test how pseudonymisation shows up in the Explorer.
- Make sure that annotation fields can be ordered.
- Add a versioning system for Explorer templates (so that it's traceable what aesthetics/fields informed what annotations).
- Add a processor that lets you merge/copy annotations from one dataset to the other. If you'd e.g. make a random filter, this way you can also add these annotations to the parent dataset.
- Allow showing media from the server if these have been donwloaded through the Download videos or Download images processors.
- All kinds of LLM features!

Hope that's all clear and everything works on your 4CAT's as well :)

dale-wahl commented:
Tested it out and seems to work fine. Tested creating and writing annotations. Sorting works well.

I ran the processor itself from the processor lists. Seems to work, but the result is odd. Looks like a filter, but does not link to a separate dataset.
![image](https://github.com/digitalmethodsinitiative/4cat/assets/32108944/cca8f24f-ad4d-4e00-bb51-82d1b5003d3e)

That's probably not how you intend on users writing annotation to the file so, if you wanted, you could remove it from the list by marking it as hidden:
https://github.com/digitalmethodsinitiative/4cat/blob/35c1c6d1e741ffd97468df79bb063e641dc497cc/processors/visualisation/download_images.py#L46C2-L46C11

dale-wahl commented:
Found a weird bug: the preview view for any processors run also include the annotation fields. E.g., run count values and preview the result to see the annotation field there too.

sal-uva commented:
That last one should definitely be fixed, thanks, I'll get to that.

The Write annotations processor works as intended; it just adds the annotations to the parent dataset (though 'Filtered' is maybe not the best word to use here).

sal-uva commented:
@dale-wahl that last bug is fixed now, @stijn-uva let me know if you want to have a look, else I'll merge 👍

stijn-uva commented on common/lib/user_input.py:
```diff
@@ -35,6 +35,8 @@ class UserInput:
     OPTION_HUE = "hue"  # colour hue
     OPTION_DATASOURCES = "datasources"  # data source toggling
+    OPTION_DATASOURCES_TABLE = "datasources_table" # a table with settings per data source
+    OPTION_DATASOURCES_TEXT = "datasources_text" # text input per data source (via dropdown)
```
`OPTION_DATASOURCES_TEXT` seems to be unused

sal-uva replied:
Removed in [4a92b14](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/4a92b1488c324674d922b8836bdc968532399307)

stijn-uva commented on datasources/douyin/search_douyin.py:
```diff
@@ -7,6 +7,7 @@
 from datetime import datetime
 
 from backend.lib.search import Search
+from common.lib.helpers import UserInput
```
Unnecessary import

stijn-uva commented on datasources/imgur/search_imgur.py:
```diff
@@ -8,8 +8,9 @@
 
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
Unnecessary import

stijn-uva commented on datasources/instagram/search_instagram.py:
```diff
@@ -10,6 +10,7 @@
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem, MissingMappedField
 from common.lib.exceptions import WorkerInterruptedException, MapItemException
+from common.lib.helpers import UserInput
```
Unnecessary import

sal-uva replied:
Removed in [c4b1943](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/c4b19434af52ecc2a57f7c8d609aa080ae5cdc15)

stijn-uva commented on datasources/linkedin/search_linkedin.py:
```diff
@@ -11,6 +11,7 @@
 
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
Unnecessary import

stijn-uva commented on datasources/parler/search_parler.py:
```diff
@@ -10,6 +10,7 @@
 
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
Unnecessary import

stijn-uva commented on datasources/tiktok/search_tiktok.py:
```diff
@@ -9,6 +9,7 @@
 
 from backend.lib.search import Search
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
Unnecessary import

stijn-uva commented on datasources/twitter-import/search_twitter.py:
```diff
@@ -9,11 +9,12 @@
 from backend.lib.search import Search
 from common.lib.helpers import strip_tags
 from common.lib.item_mapping import MappedItem
+from common.lib.helpers import UserInput
```
Unnecessary import

sal-uva commented on common/lib/config_definition.py:
```diff
@@ -305,21 +305,48 @@
     # Explorer settings
-    # The maximum allowed amount of rows (prevents timeouts and memory errors)
-    "explorer.max_posts": {
+    "explorer.__basic-explanation": {
```
Because options are sorted alphabetically!

sal-uva replied:
Settings retain their dictionary order per fcb747301809ed96f0c97491b4e2dd64fe37f295

sal-uva commented on webtool/templates/components/result-result-row.html:
```diff
@@ -17,41 +17,41 @@
+    <li>
+        <a href="{{ url_for('get_result', query_file=dataset.result_file) }}" class=" tooltip-trigger" aria-controls="tooltip-get-result-{{ dataset.key }}">
+            <i class="fas fa-download" aria-hidden="true"></i>
+            Original {{ dataset.result_file.split(".")[-1] }} ({{ dataset.get_results_path()|filesize }})
```
That is an option, but in the case of removed data files you'll get `False` -- edge case but probably a bit nicer to just use the filename's extension

stijn-uva replied:
But if the file is removed you also cannot download it, so should the button be there at all then?

sal-uva replied:
Removed it if the file doesn't exist in [3fd55ca](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/3fd55ca3840c13baac81b27f46d7c108f4cdf794)

sal-uva commented on common/lib/dataset.py:
```diff
@@ -822,13 +853,201 @@ def get_annotations(self):
+
+		# We're saving all annotations as a JSON string
+		annotations = json.dumps(annotations)
+		self.db.execute("INSERT INTO annotations(key, annotations) VALUES(%s, %s) ON CONFLICT (key) DO UPDATE SET annotations = %s ", (self.top_parent().key, annotations, annotations))
```
Fixed as of commit [868af5e](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/868af5e1e7b936e6bf67ff29d122c5ebdbe019d8)

sal-uva reviewed (commented):
Addressed in separate comments

stijn-uva commented on common/lib/helpers.py:
```diff
@@ -881,3 +898,9 @@ def folder_size(path='.'):
             total += folder_size(entry.path)
     return total
+
+def hash_values(string: str) -> str:
```
Since we are using several types of hashes in 4CAT (MD5, SHA256 and Blake2b) please indicate in the function name that this is md5. And perhaps `value` instead of `values`?

stijn-uva commented on backend/workers/expire_items.py:
```diff
@@ -62,6 +62,7 @@ def expire_datasets(self):
 				dataset = DataSet(key=dataset["key"], db=self.db)
 				if dataset.is_expired():
 					self.log.info(f"Deleting dataset {dataset.key} (expired)")
+					dataset.delete_annotations(dataset_key=dataset.key)
```
Wouldn't it make sense to add `delete_annotations()` to `delete()` instead? You probably always want to delete the annotations too if you're deleting a dataset

stijn-uva commented on common/lib/dataset.py:
```diff
@@ -1510,6 +1502,20 @@ def is_from_collector(self):
 		"""
 		return self.type.endswith("-search") or self.type.endswith("-import")
 
+	def file_exists(self):
```
Since this is just used in one place why not call `dataset.get_results_path().exists()` directly in the template?

stijn-uva commented on common/lib/helpers.py:
```diff
@@ -245,6 +246,22 @@ def convert_to_int(value, default=0):
+
+    :param value:  Value to convert
+    :param int default:  Default value, if conversion not possible
+    :return int:  Converted value
```
float? 😬

sal-uva commented:
Okay, I think this is ready to merge.

@stijn-uva maybe checking `annotation.py` is especially useful; I hope that getting/setting attributes is now done properly, but not 100% sure.

For documentation:
# Changelog
### Annotations
- An annotation can now be instantiated as an object through class `Annotation()`, which takes care of handling various annotation functions, like saving and deleting.
- The annotations table is expanded to correspond with `Annotation()` attributes, and annotations are saved as their own row in the database. The table now includes:
  - `id`: Unique ID for this annotation
  - `item_id`: ID of the item that this annotation was made for, e.g. a post ID
  - `field_id`: ID of the annotation field
  - `dataset`: Dataset key this annotation is generated from
  - `timestamp`: When this annotation was edited
  - `timestamp_created`: When this annotation was created
  - `label`: Label of annotation
  - `type`: Type of annotation (e.g. `text`)
  - `options`: Possible options
  - `value`: The actual annotation value
  - `author`: Who last edited the annotation
  - `author_original`: Who originally made the annotation
  - `by_processor`: Whether the annotation was made by a processor
  - `metadata`: Misc metadata (e.g. processor parameters)
- Processors can now easily create and save annotations by calling `self.write_annotations()` and giving it a dict of annotations. This creates and writes annotations that correspond to the parent dataset. These can then be viewed and edited in the Explorer. `Annotation()` attributes can be set explicitly, but only an `item_id` and `value` are necessary; the rest is handled in `processor.py`. This makes LLM-generated annotations easy to add. Quick example:
```
def process(self):
  annotations = []
  try:
      with self.dataset.get_results_path().open("w") as results:

          for post in self.source_dataset.iterate_items(self):

                  annotation = {"item_id": post["id"],
                                "label": "Label",
                                "value": "This is an annotation!"}
                  annotations.append(annotation)

          self.write_annotations(annotations, overwrite=self.parameters.get("overwrite", False))
          self.dataset.finish(1)
  except AnnotationException as e:
      self.dataset.finish_with_error(str(e))
```
- The annotation functions in `dataset.py` are rewritten with the `Annotation()` class and new database structure in mind.
- Error handling is improved with `AnnotationException`.
- Added a new processor that lets you download annotation metadata.
- Annotation are now deleted when datasets get deleted.
### Explorer
- The aesthetics of the Explorer are now more in line with the rest of 4CAT.
- Editing annotation fields:
  - More compact look 'n feel that makes it easy to add future features (e.g.adding default values or hiding annotations by default).
  - The naming for input fields is changed to make them less webdevvy: `dropbox` to `single choice` and `checkbox` to `multiple choice`.
  - Removed a ton of code from `explorer.js` by simply reloading the page when fields are applied. This is now also cache-less so input fields are reset.
  - Fields can now be ordered (fixes #225).
  - The user gets notified if changes to annotation fields also affect existing annotations.
- Data source templates:
  - Changed the location and structure of template files:
    - Jinja2 HTML files are now used instead of the JSON system. They are stored in `webtool/templates/explorer/datasource-templates/` .
    - CSS files are moved to `webtool/static/css/explorer/`.
  - A custom template is automatically used if template files are present (these must have the name of the datasource, e.g. `twitter.html`, and be placed in the above directories). 
  - Improved templates for Instagram, Tumblr, Twitter, and TikTok.
  - Added templates for Telegram and LinkedIn.
- Flask template filters
  - Added a filter that 'social mediafies' texts, so tags, @-mentions, and URLs get an anchor tag that leads to the original social media site.
  - Added a  filter that counts characters in a string, including emojis.
- Datasets can now be sorted by any column, as well as be reversed.
- The showing/hiding of annotations is now cached so that 'Show annotations' doesn't need to be clicked constantly.
- Annotations now have a tooltip that shows metadata on the original author and who edited it last.
- Saving is improved: you can now always save and saving is only triggered when a user hasn't edited a field for 3 seconds.
- Added a JS function that allows to display timestamps in the client's time zone.
### Data sources
- Added new items in `map_items()` for Instagram, Twitter, Telegram, LinkedIn, and TikTok, and Tumblr, corresponding to the elements that are present in the actual interface of these sites.
- Tumblr:
  - Significantly improved search:
    - Added the option to get posts by URL.
    - Allowed to mix different types of queries, like hashtag-level *and* blog-level search.
    - Added search options to add tags, reblogs with text, and reblogs with hashtags.
    - Added the option to include/exclude reblogs outside the date range.
  - Made the data source handle 'Neue Post Format' post structure.
  - Better explained the API's oddities in the search UI and the readme.
- Misc improvements and bug fixes.
### Settings
- Explorer settings has its own page on the Settings page.
- The Explorer can be enabled and disabled per data source.
- The above setting is enabled by a new `OPTION_DATASOURCES_TABLE` input. This table can easily be expanded in the future, e.g. with options to force pseudonymisation or to indicate the availability of API-callable posts outside of a dataset.
### Misc
- Removed unnecessary code and cleaned up files.
- Added a `hash_to_md5` helper function.
- Started adding some type hinting in new and existing functions.
# For future updates
- Write documentation on annotations.
- Add missing Explorer templates (e.g. Reddit).
- Allow to delete all annotations for a dataset.
- Enable setting a default value for an annotation field.
- Add a processor that lets you merge/copy annotations from one dataset to another.
  - If you'd e.g. make a random filter, this way you can also add these annotations to the parent dataset.
- Add a processor that lets you upload existing annotations
  - As csv or JSON with fields post ID --> annotation.
- Show media from the 4CAT server if these have been downloaded through the `Download videos` or `Download images` processors.
- Add LLM annotators.
  - Preferably open-source...
  - Make models selectable.
- Add statistic processors to make text analysis and coding more methodologically robust (e.g. to calculate Krippendorff's alpha).
- Add features that improve coding with multiple people (e.g. a notice that another user is currently annotating a dataset).

stijn-uva commented on common/lib/dataset.py:
```diff
@@ -517,11 +538,13 @@ def delete(self, commit=True):
 			try:
 				child = DataSet(key=child["key"], db=self.db)
 				child.delete(commit=commit)
+				child.delete_annotations()
```
This is redundant since `child.delete()` will call the same `delete_annotations()` you added below (it's recursive)

stijn-uva commented:
When trying to run this and running the migrate script to prepare 4CAT, I got this crash:

```
  Unexpected error while running migrate-1.45-1.46.py. Migration halted.
  The following exception occurred:

Traceback (most recent call last):
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/helper-scripts/migrate/migrate-1.45-1.46.py", line 113, in <module>
    dataset = db.fetchone("SELECT * FROM datasets WHERE key = '" + row["dataset"] + "';")
KeyError: 'dataset'
```

stijn-uva commented on webtool/templates/explorer/controls.html:
```diff
@@ -0,0 +1,69 @@
+						<li class="annotation-control-button">
+							<a class="button-like-small" id="toggle-annotation-fields"><i class="fas fa-edit"></i> Edit fields</a>
+						</li>
+						<li class="annotation-control-button"><a class="button-like-small{% if not annotation_fields %} disabled{% endif %}" id="toggle-annotations"><i class="fas fa-eye"></i> Show annotations</a></li>
```
These buttons can be toggled even when they are disabled

stijn-uva commented on common/lib/dataset.py:
```diff
@@ -377,10 +376,32 @@ def iterate_items(self, processor=None, warn_unmappable=True, map_missing="defau
+			if annotation_labels:
+
+				# Get annotations for this specific post
+				post_annotations = self.get_annotations(item_id=mapped_item.data["id"])
```
I get a crash here after adding a multiple-choice annotation field to a dataset:

```
ERROR:webtool:Exception on /results/74a03678e6a669f8db441ac431f31de2/explorer/ [GET]
Traceback (most recent call last):
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask_limiter/extension.py", line 544, in __inner
    return obj(*a, **k)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask_login/utils.py", line 290, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/webtool/lib/helpers.py", line 374, in decorated_view
    return func(*args, **kwargs)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/webtool/views/views_explorer.py", line 97, in explorer_dataset
    for row in dataset.iterate_items(warn_unmappable=False):
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/common/lib/dataset.py", line 386, in iterate_items
    post_annotations = self.get_annotations(item_id=mapped_item.data["id"])
AttributeError: 'dict' object has no attribute 'data'

```

sal-uva replied:
Fixed in [55d81f5](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/55d81f5217feb6380b6ff014bb67384401833fb1)

stijn-uva reviewed (changes requested):
This is as far as I got!

stijn-uva reviewed (commented):
Got an error while trying to save annotations:

```
ERROR:webtool:Exception on /explorer/save_annotations/a565f95306bcf5182adc023aa82a8d59 [POST]
Traceback (most recent call last):
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask_limiter/extension.py", line 544, in __inner
    return obj(*a, **k)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/flask_login/utils.py", line 290, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/webtool/lib/helpers.py", line 332, in decorated_view
    return func(*args, **kwargs)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/webtool/lib/helpers.py", line 332, in decorated_view
    return func(*args, **kwargs)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/webtool/views/views_explorer.py", line 214, in explorer_save_annotations
    annotations_saved = dataset.save_annotations(annotations, overwrite=True)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/common/lib/dataset.py", line 1787, in save_annotations
    annotation = Annotation(data=annotation_data, db=self.db)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/common/lib/annotation.py", line 148, in __init__
    self.write_to_db()
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/common/lib/annotation.py", line 223, in write_to_db
    return self.db.upsert("annotations", data=db_data, constraints=["label", "dataset", "item_id"])
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/common/lib/database.py", line 270, in upsert
    cursor.execute(query, replacements)
  File "/Users/stijn/surfdrive/PycharmProjects/4cat/venv/lib/python3.9/site-packages/psycopg2/extras.py", line 236, in execute
    return super().execute(query, vars)
psycopg2.errors.InvalidColumnReference: there is no unique or exclusion constraint matching the ON CONFLICT specification
```

Some other things I noted:

- I kind of liked the always-in-view toolbar in the old annotations layout, so I could scroll down the page and then open annotations, instead of having to scroll back up to do so
- The 'save annotations' button saves so quickly that it's sometimes hard to know if the annotations were actually saved. Maybe a 'Annotations saved!' tooltip-like notice that fades away after a second would be useful
- Maybe annotation fields should be shown by default after saving/editing fields? You would usually want to see them anyway. I could even see a case for not hiding them, ever, and just making them always visible (since annotation is one of the main purposes of the explorer view anyway)

sal-uva commented on processors/machine_learning/gpt.py:
```diff
@@ -0,0 +1,279 @@
+		except ValueError:
+			self.dataset.finish_with_error("Max tokens must be a number")
+
+		self.dataset.delete_parameter("api_key")  # sensitive, delete after use
```
resolved in [a1bebe8](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/a1bebe8bdd52585532b72d607fe62f15e7c81fab)

sal-uva commented on datasources/instagram/search_instagram.py:
```diff
@@ -218,6 +222,14 @@ def parse_itemlist_item(node):
+        # Instagram posts also allow 'Collabs' with up to one co-author
+        coauthor = {"coauthor": "", "coauthor_fullname": "", "coauthor_id": ""}
+        if node.get("coauthor_producers"):
+            coauthor_node = node["coauthor_producers"][0]
```
resolved in [6bec107](https://github.com/digitalmethodsinitiative/4cat/pull/428/commits/6bec1076399c75e00f9bb8c4e7074b299e4f8f2d)

sal-uva commented:
@stijn-uva I think this is once again mergeable. I addressed the above comments and tested migration with an 'old' database. If you could double-check whether the migration also goes smooth for an old database *with annotations*, that'd be great.

To do: add more templates!

dale-wahl commented on webtool/views/views_explorer.py:
```diff
@@ -0,0 +1,392 @@
+	# Else return the amount of fields saved.
+	return str(annotations_saved)
+
+def sort_and_iterate_items(dataset: DataSet, sort="", reverse=False, **kwargs) -> dict:
```
This is a very heavy load

dale-wahl commented:
Still reviewing, but I had some thoughts about the `sort_and_iterate_items` function after watching my RAM get eaten on a larger dataset and added a different implementation to `DataSet` itself in case we want to use it elsewhere (the Explorer is not using it yet though I tested it there and it was fine). The idea being to reduce memory particularly as we may have multiple users of the Explorer at once. I noticed after testing my implementation though that you call the sort function for every page. I think we should cache the results (using my implementation, we could just keep the file and use that unless the sort is changed). Because the sort is so intense, we might only run it if a "Sort" button is pushed, particularly because it will run immediately and you may want to change both what is sorted and whether it is reversed. 

Also fixed up the migrate script after it dropped my annotations. `migrate.py` is run twice often. Always in Docker and potentially in other cases (e.g. if `migrate.py` for some reasons hangs or fails) so a check is best.

sal-uva commented:
> Still reviewing, but I had some thoughts about the `sort_and_iterate_items` function after watching my RAM get eaten on a larger dataset and added a different implementation to `DataSet` itself in case we want to use it elsewhere (the Explorer is not using it yet though I tested it there and it was fine). The idea being to reduce memory particularly as we may have multiple users of the Explorer at once. I noticed after testing my implementation though that you call the sort function for every page. I think we should cache the results (using my implementation, we could just keep the file and use that unless the sort is changed). Because the sort is so intense, we might only run it if a "Sort" button is pushed, particularly because it will run immediately and you may want to change both what is sorted and whether it is reversed.
> 
> Also fixed up the migrate script after it dropped my annotations. `migrate.py` is run twice often. Always in Docker and potentially in other cases (e.g. if `migrate.py` for some reasons hangs or fails) so a check is best.

Thanks @dale-wahl. I tested out your new sorting function and it worked nicely.

To merge or not to merge?

sal-uva commented:
> I figured out a way to speed up the chunk sorting and even went so far as making temp files to allow your pagination to not require a new sort each time... but ultimately it is way too slow. Nginx will timeout before it can load so I decided it was not worth it to push those changes. Still using the chunk sorting prevents running out of memory and crashing the whole thing on big datasets. I would recommend disabling sort for datasets larger than your `explorer.max_posts` setting so users do not tax the servers only to get 500 errors. If a user needs to sort datasets that large, we can make a processor using the DataSet sort function pretty easily now.
> 
> Other than that, I did some tests on the explorer function and found no other obvious issues. You do have a number of changes to datasets and processors that I did not test. They did not seem related to the Explorer for the most part anyway.

Thanks a lot for the work, Dale. I guess the sorting issue pertains to a larger question on what 4CAT is/should be, including whether datasets should be in a database...

I will disable sorting for large datasets and merge!
